### PR TITLE
Allow signatures via a Kinto plugin

### PR DESCRIPTION
**Pull request by almet**
Adds the ability to trigger the signature of the data from an HTTP request.

Consider two Kinto instances: "A, the authoritative" and "S, the signer".

1. When S starts-up, it retrieves all the records from A and stores the "last_modified" value of the collection.
2. A client sends updates to S. When all the data has been sent, a specific field is updated on the collection to ask the signer to sign the collection.
3. The signer gets all its local data (of the collection), computes a hash and generates a signature out of it.
4. The signature + the modified elements are sent to A.
5. Then, the signer updates its current collection status back to "signed" (with the signature and hash).

In order to test the signature flow, you can run the functional tests:

```bash
$ make run-remote
$ make run-signer
$ make functional
```

Stuff left to do:

- [x] Create a hook that makes it possible to trigger a signature on the server;
- [x] Add functional tests checking this works propertly (with the replication from one server to the other);
- [ ] Update the current tests so that the component is unit tested;

### CODE REVIEW

**Comment by almet** [comment]
I believe this is about to be ready.

I want to add some more features there, among them:

- [ ] The ability to configure which backend to use from the configuration;
- [ ] The ability to have a remote bucket/collection that differs from the local ones;
- [ ] Maybe allow to have multiple tuple of bucket/collection that can be signed by one origin server.
- [ ] Add a way to gather the remote changes from an origin (what was in the `gather_changes.py` file)

But I would like to address these as enhancements, and land this pull request as soon as possible, it's been too long already.

@natim @leplatrem if you have some time to review, that would be much appreciated. Thanks.

File: README.rst [hunk 1]
```diff
@@ -10,32 +10,93 @@ Kinto updater
...
+contained in configured bucket/collection of the origin.
+
+Triggering a signature on the authority
+=======================================
+
+Once started, the authority is behaving like a normal Kinto server, until you
+ask for a signature of the collection. To trigger this signature operation,
+you need to add a specific field on the **collection**: ``status:to-sign``.
+
+Here is how to do it with `httpie`:
```
Comment (Natim) [inline README.rst]: ``httpie``

File: README.rst [hunk 2]
```diff
@@ -10,32 +10,93 @@ Kinto updater
...
+
+.. code-block::
+
+  echo '{"data": {"status": "to-sign"}}' | http PATCH http://0.0.0.0:8888/v1/buckets/default/collections/tasks --auth user:pass
```
Comment (Natim) [inline README.rst]: Shouldn't we have 4 spaces before echo ?

File: README.rst [hunk 3]
```diff
@@ -10,32 +10,93 @@ Kinto updater
...
+This works with two Kinto instances:
 
-  2a. check the validity of the signature (with the pub key)
-  2b. compute the collection hash locally;
-  2c. compare it with the one stored in the collection;
+- **A, the authority** (also known as "the signer"). It is where the original
+  data are sent. The authority is configured to sign the data to a specific
+  "origin".
+- **O, the origin**, which will end up distributing the data and the signatures.
+  It is where the data is retrieved.
```
Comment (leplatrem) [inline README.rst]: It is where the client retrieve the data ?

File: README.rst [hunk 4]
```diff
@@ -10,32 +10,93 @@ Kinto updater
...
-1. Gather the latest changes from the server,
-2. Check their validity
+This works with two Kinto instances:
 
-  2a. check the validity of the signature (with the pub key)
-  2b. compute the collection hash locally;
-  2c. compare it with the one stored in the collection;
+- **A, the authority** (also known as "the signer"). It is where the original
+  data are sent. The authority is configured to sign the data to a specific
+  "origin".
```
Comment (leplatrem) [inline README.rst]: *for a specific origin* ?

File: README.rst [hunk 5]
```diff
@@ -10,32 +10,93 @@ Kinto updater
 What does this do?
 ==================
 
-This will:
+**Kinto signer** is a `kinto <https://kinto.readthedocs.org>`_ plugin that
+makes it possible so sign the updates of kinto collections. In other words,
+it's a way to verify that the data the client has got is the data the original
```
Comment (leplatrem) [inline README.rst]: that the data retrieved by the client are the data

File: README.rst [hunk 6]
```diff
@@ -10,32 +10,93 @@ Kinto updater
 What does this do?
 ==================
 
-This will:
+**Kinto signer** is a `kinto <https://kinto.readthedocs.org>`_ plugin that
+makes it possible so sign the updates of kinto collections. In other words,
```
Comment (leplatrem) [inline README.rst]: typo: possible to sign

nitto: Kinto

File: README.rst [hunk 7]
```diff
@@ -10,32 +10,93 @@ Kinto updater
 What does this do?
 ==================
 
-This will:
+**Kinto signer** is a `kinto <https://kinto.readthedocs.org>`_ plugin that
```
Comment (leplatrem) [inline README.rst]: nit: Kinto

File: README.rst [hunk 8]
```diff
@@ -10,32 +10,93 @@ Kinto updater
...
+.. image::
+   schema.png
 
-How to install?
-===============
+Configuring kinto-signer
+========================
 
-::
+To install this plugin in a kinto server, a few configuration variables needs
```
Comment (leplatrem) [inline README.rst]: nitto Kinto

Comment (leplatrem) [inline README.rst]: variables need to be set

File: README.rst [hunk 9]
```diff
@@ -10,32 +10,93 @@ Kinto updater
...
++=================================+==========================================================================+
+| kinto.signer.remote_server_url  | The complete location of the remote server URL (the origin)              |
+|                                 | For instance ``https://localhost:8000/v1/``                              |
++---------------------------------+--------------------------------------------------------------------------+
+| kinto.signer.bucket             | The name of the bucket on which signatures can be applied.               |
+|                                 | *Current limitation: the bucket should be the same locally and remotely* |
++---------------------------------+--------------------------------------------------------------------------+
+| kinto.signer.collection         | The name of the collection on which signatures can be applied            |
+|                                 | *Current limitation: the collection should be the same locally and       |
+|                                 | remotely                                                                 |
```
Comment (leplatrem) [inline README.rst]: missing trailing *

File: README.rst [hunk 10]
```diff
@@ -10,32 +10,93 @@ Kinto updater
...
 
-  $ mkvirtualenv kinto-updater
-  $ pip install -r requirements.txt
++---------------------------------+--------------------------------------------------------------------------+
+| Setting name                    | What does it do?                                                         |
++=================================+==========================================================================+
+| kinto.signer.remote_server_url  | The complete location of the remote server URL (the origin)              |
+|                                 | For instance ``https://localhost:8000/v1/``                              |
++---------------------------------+--------------------------------------------------------------------------+
+| kinto.signer.bucket             | The name of the bucket on which signatures can be applied.               |
```
Comment (leplatrem) [inline README.rst]: why *can* be applied and not *will be applied* ?

File: README.rst [hunk 11]
```diff
@@ -10,32 +10,93 @@ Kinto updater
...
+|                                 | *Current limitation: the bucket should be the same locally and remotely* |
++---------------------------------+--------------------------------------------------------------------------+
+| kinto.signer.collection         | The name of the collection on which signatures can be applied            |
+|                                 | *Current limitation: the collection should be the same locally and       |
+|                                 | remotely                                                                 |
++---------------------------------+--------------------------------------------------------------------------+
+| kinto.signer.remote_server_auth | The authentication to use on the remote server. Should be specified as   |
+|                                 | "user:password".                                                         |
++---------------------------------+--------------------------------------------------------------------------+
+| kinto.signer.private_key        | The location of the ECDSA private key to use to apply the signatures.    |
```
Comment (leplatrem) [inline README.rst]: location *on disk* ? Absolute, relative path ?

File: README.rst [hunk 12]
```diff
@@ -10,32 +10,93 @@ Kinto updater
...
+  kinto_signer.bucket = buck
+  kinto_signer.collection = coll
+  kinto_signer.remote_server_url = http://localhost:7777/v1
+  kinto_signer.remote_server_auth = user:p4ssw0rd
+  kinto_signer.private_key = kinto_signer/tests/config/ecdsa.private.pem
 
+Starting the authority for the first time
+=========================================
+
+Each time the authority starts, it will reach the origin and gets all the data
```
Comment (leplatrem) [inline README.rst]: nit: and get

File: README.rst (hunk 10 above)
Comment (leplatrem) [inline README.rst]: ok got it (saw below)

File: README.rst [hunk 13]
```diff
@@ -10,32 +10,93 @@ Kinto updater
...
+
+  echo '{"data": {"status": "to-sign"}}' | http PATCH http://0.0.0.0:8888/v1/buckets/default/collections/tasks --auth user:pass
+
+From there, the authority will:
+
+1. Retrieve all records on the collection, compute a hash of the records, and
+   generate a signature out of it.
+2. Send all local changes to the Origin server, **with a signature**.
+3. Update the collection metadata with ``status:signed``.
+
```
Comment (leplatrem) [inline README.rst]: IMO these three paragraphs should be put on top, before configuration.

File: kinto.ini [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+multiauth.policies = basicauth
+
+kinto.includes = kinto_signer.hook
+
+; Kinto Updater configuration.
+; XXX Put this in a section instead.
+kinto_signer.bucket =
+kinto_signer.collection =
+kinto_signer.remote_server_url = http://localhost:7777
+kinto_signer.private_key = 'test.pem'
```
Comment (leplatrem) [inline kinto.ini]: Above you wrote `kinto.signer.bucket`

File: kinto_signer/generate_keypair.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+    private_key, public_key = backend.generate_keypair()
+
+    with open(private_key_location, 'wc') as tmp_file:
+        tmp_file.write(private_key)
+
+    with open(public_key_location, 'wc') as tmp_file:
+        tmp_file.write(public_key)
+
+
+if __name__ == "__main__":  # pragma: no cover
```
Comment (leplatrem) [inline kinto_signer/generate_keypair.py]: nit: this script deserves a section in docs :)

File: README.rst [hunk 16]
```diff
@@ -10,32 +10,93 @@ Kinto updater
...
-::
+To install this plugin in a kinto server, a few configuration variables needs
+to be set.
 
-  $ mkvirtualenv kinto-updater
-  $ pip install -r requirements.txt
++---------------------------------+--------------------------------------------------------------------------+
+| Setting name                    | What does it do?                                                         |
++=================================+==========================================================================+
+| kinto.signer.remote_server_url  | The complete location of the remote server URL (the origin)              |
```
Comment (leplatrem) [inline README.rst]: in code you read `kinto_signer`

File: kinto_signer/hook.py [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+        payload = event.payload
+        resource_name = payload['resource_name']
+        action = payload['action']
+
+        # XXX Replace the filtering by events predicates (on the next Kinto
+        # release)
+        # XXX Add a concept of local and remote buckets/collections
+        correct_bucket = True  # payload.get('bucket_id') == expected_bucket
+        correct_coll = payload.get('collection_id') == expected_collection
+        is_coll = resource_name == 'collection'
```
Comment (leplatrem) [inline kinto_signer/hook.py]: In Kinto >= 1.11 you can filter during subcription
http://cliquet.readthedocs.org/en/latest/reference/notifications.html#filtering

File: README.rst [hunk 18]
```diff
@@ -10,32 +10,93 @@ Kinto updater
...
+
+Each time the authority starts, it will reach the origin and gets all the data
+contained in configured bucket/collection of the origin.
+
+Triggering a signature on the authority
+=======================================
+
+Once started, the authority is behaving like a normal Kinto server, until you
+ask for a signature of the collection. To trigger this signature operation,
+you need to add a specific field on the **collection**: ``status:to-sign``.
```
Comment (leplatrem) [inline README.rst]: nit: ``status: "to-sign"``

File: kinto_signer/hook.py [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+        is_creation = action in ('create', 'update')
+
+        if correct_coll and correct_bucket:
+            if is_creation and is_coll:
```
Comment (leplatrem) [inline kinto_signer/hook.py]: I suggest you return if condition is negative (save levels of indentation)

File: kinto_signer/signer.py [hunk 20]
```diff
@@ -1,22 +1,28 @@
...
+        return pem
+
+    def export_public_key_as_pem(self, key):
+        return key.public_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PublicFormat.SubjectPublicKeyInfo
+        )
 
     def load_private_key(self):
         # Check settings validity
```
Comment (leplatrem) [inline kinto_signer/signer.py]: you can raise `pyramid.expections.ConfigurationError` instead of ValueError

File: kinto_signer/signer.py [hunk 21]
```diff
@@ -29,34 +35,61 @@ def load_private_key(self):
...
+
+class RSABackend(SignerBackend):
+    """Local RSA signature backend.
+    """
+    key_size = 4096
+
+    def _get_padding(self):
+        return padding.PSS(
+            mgf=padding.MGF1(hashes.SHA256()),
+            salt_length=padding.PSS.MAX_LENGTH)
```
Comment (leplatrem) [inline kinto_signer/signer.py]: 这是中国对我来说

File: kinto_signer/tests/functional.py [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+        resp = requests.post(flush_url)
+        resp.raise_for_status()
+
+    def get_user_id(self, credentials):
+        hmac_secret = self.signer_config.get(
+            'app:main',
+            'cliquet.userid_hmac_secret')
+        credentials = '%s:%s' % credentials
+        digest = cliquet_utils.hmac_digest(hmac_secret, credentials)
+        return 'basicauth:%s' % digest
```
Comment (leplatrem) [inline kinto_signer/tests/functional.py]: Why not `GET /` ?

File: kinto_signer/tests/functional.py [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+                batch.create_record(data={'foo': 'bar', 'n': n})
+
+        # Replicate the remote data locally:
+        origin = Client(
+            server_url=self.remote_url,
+            auth=self.auth,
+            bucket='buck',
+            collection='coll'
+        )
+        destination = Client(
```
Comment (leplatrem) [inline kinto_signer/tests/functional.py]: nit: rename authority

File: kinto_signer/tests/functional.py [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+        # Ensure the remote data is signed properly.
+        data = self.remote_client.get_collection()
+        signature = data['data']['signature']
+        assert signature is not None
+
+        records = self.remote_client.get_records()
+        assert len(records) == 15
+        local_hash = compute_hash(records)
+        self.signer.verify(local_hash, signature)
+
```
Comment (leplatrem) [inline kinto_signer/tests/functional.py]: niiiiice!

File: kinto_signer/tests/test_updater.py [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+        batch.__enter__().patch_collection.assert_called_with(
+            data={'signature': 'signature'})
+
+    def test_sign_and_update_remote(self):
+        records = [{'id': idx, 'foo': 'bar %s' % idx}
+                   for idx in range(1, 3)]
+        self.storage.get_all.return_value = (records, 2)
+        self.updater.update_remote = mock.MagicMock()
+
+        self.updater.sign_and_update_remote()
```
Comment (leplatrem) [inline kinto_signer/tests/test_updater.py]: missing assert ?

File: kinto_signer/updater.py [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+                                      COMPARISON.GT)
+            storage_kwargs['filters'] = [gt_last_modified, ]
+
+        parent_id = "/buckets/%s/collections/%s" % (
+            self.bucket_id, self.collection_id)
+
+        records, _ = self.storage.get_all(
+            parent_id=parent_id,
+            collection_id='record', **storage_kwargs)
+        return records
```
Comment (leplatrem) [inline kinto_signer/updater.py]: We are starting this kind of code in every plugin :) Would be nice to initiative a reflexion once we have a couple of stabilized plugins

File: kinto_signer/updater.py [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+
+    def get_remote_last_modified(self):
+        endpoint = self.remote._get_endpoint('records')
+        # XXX Replace with a HEAD request.
+        _, headers = self.remote.session.request('get', endpoint)
+        return int(headers['Etag'].strip('"'))
```
Comment (leplatrem) [inline kinto_signer/updater.py]: `ETag`

File: kinto_signer/updater.py [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+
+    def update_remote(self, new_hash, signature):
+        last_modified = self.get_remote_last_modified()
+        new_records = self.get_collection_records(last_modified)
+
+        # Update the remote collection.
+        with self.remote.batch() as batch:
+            for record in new_records:
+                batch.update_record(data=record, id=record['id'], safe=False)
+            batch.patch_collection(data={'signature': signature})
```
Comment (leplatrem) [inline kinto_signer/updater.py]: > note: We definitely need to have an implicit retry on batch

File: kinto_signer/updater.py [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+        2. Compute a hash of these records;
+        3. Ask the signer for a signature;
+        4. Send all records since the last_modified field of the Authoritative
+           server;
+        5. Send the signature to the Authoritative server.
+        """
+        records = self.get_collection_records()
+        new_hash = hasher.compute_hash(records)
+        signature = self.signer.sign(new_hash)
+        self.update_remote(new_hash, signature)
```
Comment (leplatrem) [inline kinto_signer/updater.py]: nit: move on top of class?

**Comment by leplatrem** [comment]
Nice piece of work!

I didn't try to run it locally.

Beware: I wrote every comment that came in my mind :) And didn't write any compliment :relaxed:

File: kinto_signer/signer.py (hunk 20 above)
Comment (almet) [inline kinto_signer/signer.py]: I actually like the idea of having the code of the signer completely separate from pyramid.

File: kinto_signer/tests/functional.py (hunk 22 above)
Comment (almet) [inline kinto_signer/tests/functional.py]: That avoids a roundtrip to the server I guess?

File: kinto_signer/tests/test_updater.py (hunk 25 above)
Comment (almet) [inline kinto_signer/tests/test_updater.py]: I'm actually not sure about what to test here. The sign_and_update_remote is just calling other functions that were already tested.

File: kinto_signer/updater.py (hunk 26 above)
Comment (almet) [inline kinto_signer/updater.py]: That's exactly what I was telling to @natim :-D

File: kinto_signer/updater.py (hunk 28 above)
Comment (almet) [inline kinto_signer/updater.py]: Referencing https://github.com/Kinto/kinto.py/issues/39 :-)

**Comment by almet** [comment]
that's cool, thanks! updated with your changes.

File: kinto_signer/tests/functional.py (hunk 22 above)
Comment (leplatrem) [inline kinto_signer/tests/functional.py]: hum, but I don't think that duplicating the construction of user-id is a viable option though

File: kinto_signer/signer.py (hunk 20 above)
Comment (leplatrem) [inline kinto_signer/signer.py]: oh my bad!

File: kinto_signer/hook.py (hunk 17 above)
Comment (leplatrem) [inline kinto_signer/hook.py]: @almet I believe it's better to use those `for_actions=()` and `for_resources=()` predicates

File: kinto_signer/tests/functional.py (hunk 22 above)
Comment (almet) [inline kinto_signer/tests/functional.py]: For later reference: this is actually in the functional tests, so we agreed that it's okay.

**Comment by almet** [comment]
Waiting for travis to build, and then we'll be good to go!

**Pull request merged**